cmd/dmsgpost/dmsgpost.go

```diff
@@ -0,0 +1,226 @@
+		if dmsgpostLog == nil {
+			dmsgpostLog = logging.MustGetLogger("dmsgpost")
+		}
+		if logLvl != "" {
+			if lvl, err := logging.LevelFromString(logLvl); err == nil {
+				logging.SetLevel(lvl)
+			}
+		}
+
+		ctx, cancel := cmdutil.SignalContext(context.Background(), dmsgpostLog)
+		defer cancel()
+
+		pk, err := sk.PubKey()
+		if err != nil {
+			pk, sk = cipher.GenerateKeyPair()
+		}
+
+		u, err := parseURL(args)
+		if err != nil {
+			dmsgpostLog.WithError(err).Fatal("failed to parse provided URL")
+		}
+
+		dmsgC, closeDmsg, err := startDmsg(ctx, pk, sk)
+		if err != nil {
+			dmsgpostLog.WithError(err).Fatal("failed to start dmsg")
+		}
+		defer closeDmsg()
+
+		httpC := http.Client{Transport: dmsghttp.MakeHTTPTransport(ctx, dmsgC)}
+
+		req, err := http.NewRequest(http.MethodPost, u.URL.String(), strings.NewReader(dmsgpostData))
+		if err != nil {
+			dmsgpostLog.WithError(err).Fatal("Failed to formulate HTTP request.")
+		}
+		req.Header.Set("Content-Type", "text/plain")
+
+		resp, err := httpC.Do(req)
+		if err != nil {
+			dmsgpostLog.WithError(err).Fatal("Failed to execute HTTP request.")
+		}
+
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				dmsgpostLog.WithError(err).Fatal("Failed to close response body")
+			}
+		}()
+		respBody, err := io.ReadAll(resp.Body)
+		if err != nil {
+			dmsgpostLog.WithError(err).Fatal("Failed to read respose body.")
+		}
+		fmt.Println(string(respBody))
+	},
+}
+
+// URL represents a dmsg http URL.
+type URL struct {
+	dmsg.Addr
+	url.URL
+}
+
+// Fill fills the internal fields from an URL string.
+func (du *URL) fill(str string) error {
+	u, err := url.Parse(str)
+	if err != nil {
+		return err
+	}
+
+	if u.Scheme == "" {
+		return errors.New("URL is missing a scheme")
+	}
+
+	if u.Host == "" {
+		return errors.New("URL is missing a host")
+	}
+
+	du.URL = *u
+	return du.Addr.Set(u.Host)
+}
+
+func parseURL(args []string) (*URL, error) {
+	if len(args) == 0 {
+		return nil, errors.New("no URL(s) provided")
+	}
+
+	if len(args) > 1 {
+		return nil, errors.New("multiple URLs is not yet supported")
+	}
+
+	var out URL
+	if err := out.fill(args[0]); err != nil {
+		return nil, fmt.Errorf("provided URL is invalid: %w", err)
+	}
+
+	return &out, nil
+}
+
+func startDmsg(ctx context.Context, pk cipher.PubKey, sk cipher.SecKey) (dmsgC *dmsg.Client, stop func(), err error) {
+	dmsgC = dmsg.NewClient(pk, sk, disc.NewHTTP(dmsgDisc, &http.Client{}, dmsgpostLog), &dmsg.Config{MinSessions: dmsgSessions})
+	go dmsgC.Serve(context.Background())
+
+	stop = func() {
+		err := dmsgC.Close()
+		dmsgpostLog.WithError(err).Debug("Disconnected from dmsg network.")
+		fmt.Printf("\n")
+	}
+	dmsgpostLog.WithField("public_key", pk.String()).WithField("dmsg_disc", dmsgDisc).
+		Debug("Connecting to dmsg network...")
+
+	select {
+	case <-ctx.Done():
+		stop()
+		return nil, nil, ctx.Err()
+
+	case <-dmsgC.Ready():
+		dmsgpostLog.Debug("Dmsg network ready.")
+		return dmsgC, stop, nil
+	}
+}
```

Comment:
Do we need this logic? I assumed we had a way inside of dmsg to instantiate a new client and would have a method to call on said client instead of reimplementing this logic here.

Reply:
Previously I had copied all the logic into the dmsgget because I wasn't able to change it to work correctly.
I believe what I was trying to do was to make dmsgget able to output to stdout, but the progress writer was not implemented with the logging that the rest of the program uses. So basically I had to copy it there in order to silence that if the output was stdout

but generally you would be correct in your statement.

So everything you are citing here as issues reflects the current state of dmsgget. I simply copied that implementation and modified it to do POST instead of GET